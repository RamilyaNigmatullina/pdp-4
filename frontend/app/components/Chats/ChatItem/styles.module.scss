@import "~stylesheets/colors";

.active {
  background: $grey-light-10;
}

.chatItemContainer {
  clear: both;
  overflow: hidden;
}

.chatItem {
  border-bottom: 1px solid $grey-dark-20;
  cursor: pointer;
  margin: 0;
  min-height: 76px;
  padding: 18px 16px 10px;

  &:hover {
    background: $grey;
  }
}

.details {
  display: flex;
  line-height: 1.3;
}

.interlocutorFullName {
  color: $grey-dark-65;
  font-size: 15px;
  margin: 0 0 8px 0;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  width: 100%;
}

.date {
  color: $grey-dark-65;
  float: right;
  font-size: 15px;
  padding-left: 3px;
  white-space: nowrap;
}

.badge {
  float: right;
  font-size: 12px;
}

.interlocutorAvatar {
  float: left;
  margin-right: 8px;
}

.lastMessage {
  color: $grey-dark-35;
  font-size: 14px;
  margin: auto;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  width: 100%;
}
