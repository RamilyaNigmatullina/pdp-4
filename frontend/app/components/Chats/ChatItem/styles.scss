@import "~bootstrap/scss/functions";
@import "~bootstrap/scss/variables";
@import "~bootstrap/scss/mixins/breakpoints";

.chats-list {
  flex-direction: column;
  width: 350px;

  @include media-breakpoint-down(md) {
    display: none;
  }
}

.chat-link {
  &,
  &:hover,
  &:active,
  &:focus,
  &:visited {
    color: black;
    text-decoration: none;
  }
}

.last-message {
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.chat_people {
  clear: both;
  overflow: hidden;
}

.chat_list {
  border-bottom: 1px solid #c4c4c4;
  cursor: pointer;
  margin: 0;
  padding: 18px 16px 10px;

  &:hover {
    background: #f7f7f7;
  }
}

.inbox_chat {
  height: 550px;
  overflow-y: scroll;
}

.chat_ib h5 {
  color: #464646;
  font-size: 15px;
  margin: 0 0 8px 0;
}

.chat_ib h5 span {
  float: right;
  font-size: 13px;
}

.chat_ib p {
  color: #989898;
  font-size: 14px;
  margin: auto;
}

.chat_img {
  float: left;
  width: 11%;
}

.chat_ib {
  float: left;
  padding: 0 0 0 15px;
  width: 88%;
}

.active_chat {
  background: #ebebeb;
}
