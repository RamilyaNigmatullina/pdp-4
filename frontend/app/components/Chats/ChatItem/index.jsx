import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import moment from 'moment';
import styles from './styles.module.scss';

class ChatItem extends React.Component {
  renderAvatar = () => (
    <div className={styles.interlocutorAvatar}>
      <img
        className={classNames(styles.image, 'rounded-circle')}
        height="32"
        src={this.props.chat.interlocutor.avatar}
        width="32"
      />
    </div>
  );

  renderDate = () => (
    <div className={styles.date}>
      { moment(this.props.chat.last_message.created_at).calendar() }
    </div>
  );

  renderFullNameAndDate = () => {
    const { chat } = this.props;

    return (
      <div className={styles.details}>
        <div className={styles.interlocutorFullName}>
          { chat.interlocutor.full_name }
        </div>
        { chat.last_message && this.renderDate() }
      </div>
    );
  }

  renderUnreadMessagesBadge = () => (
    <div className={classNames(styles.badge, 'badge badge-pill badge-secondary')}>
      { this.props.chat.unread_messages_count }
    </div>
  );

  renderLastMessageAndUnreadMessagesBadge = () => {
    const { chat } = this.props;

    return (
      <div className={styles.details}>
        <div className={styles.lastMessage}>
          { chat.last_message && chat.last_message.text }
        </div>
        { !!chat.unread_messages_count && this.renderUnreadMessagesBadge() }
      </div>
    );
  }

  render() {
    const { chat, onClick, isCurrentChat } = this.props;
    const chatItemClasses = classNames(styles.chatItem, { [styles.active]: isCurrentChat });

    return (
      <div className={chatItemClasses} onClick={(e) => onClick(e, chat)}>
        <div className={styles.chatItemContainer}>
          { this.renderAvatar() }
          { this.renderFullNameAndDate() }
          { this.renderLastMessageAndUnreadMessagesBadge() }
        </div>
      </div>
    );
  }
}

ChatItem.propTypes = {
  isCurrentChat: PropTypes.bool.isRequired,
  chat: PropTypes.object.isRequired,
  onClick: PropTypes.func.isRequired,
};

export default ChatItem;
