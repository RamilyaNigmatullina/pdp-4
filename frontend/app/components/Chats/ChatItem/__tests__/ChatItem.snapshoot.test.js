import React from 'react';
import { shallow } from 'enzyme';
import ChatItem from '../index';

describe('ChatItem', () => {
  let props;

  const render = () => shallow(<ChatItem {...props} />);

  beforeEach(() => {
    props = {
      isCurrentChat: false,
      chat: {
        id: 1,
        unread_messages_count: 0,
        interlocutor: {
          id: 1,
          full_name: 'John Smith',
          avatar: '/john_smith_avatar.jpg',
        },
        last_message: {
          id: 1,
          text: 'Hello! How are you?',
          sender_id: 1,
          created_at: '2020-03-15T17:39:18.327Z',
        },
      },
      onClick: jest.fn(),
    };
  });

  it('renders chat item', () => {
    const component = render();

    expect(component).toMatchSnapshot();
  });

  describe('with unread messages', () => {
    beforeEach(() => {
      props.chat.unread_messages_count = 2;
    });

    it('renders chat item', () => {
      const component = render();

      expect(component).toMatchSnapshot();
    });
  });

  describe('with current chat', () => {
    beforeEach(() => {
      props.isCurrentChat = true;
    });

    it('renders chat item', () => {
      const component = render();

      expect(component).toMatchSnapshot();
    });
  });
});
