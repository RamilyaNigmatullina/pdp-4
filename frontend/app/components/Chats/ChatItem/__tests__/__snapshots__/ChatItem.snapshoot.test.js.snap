// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatItem renders chat item 1`] = `
<div
  className="chatItem"
  onClick={[Function]}
>
  <div
    className="chatItemContainer"
  >
    <div
      className="interlocutorAvatar"
    >
      <img
        className="rounded-circle"
        height="32"
        src="/john_smith_avatar.jpg"
        width="32"
      />
    </div>
    <div
      className="details"
    >
      <div
        className="interlocutorFullName"
      >
        John Smith
      </div>
      <div
        className="date"
      >
        Today at 8:39 PM
      </div>
    </div>
    <div
      className="details"
    >
      <div
        className="lastMessage"
      >
        Hello! How are you?
      </div>
    </div>
  </div>
</div>
`;

exports[`ChatItem with current chat renders chat item 1`] = `
<div
  className="chatItem active"
  onClick={[Function]}
>
  <div
    className="chatItemContainer"
  >
    <div
      className="interlocutorAvatar"
    >
      <img
        className="rounded-circle"
        height="32"
        src="/john_smith_avatar.jpg"
        width="32"
      />
    </div>
    <div
      className="details"
    >
      <div
        className="interlocutorFullName"
      >
        John Smith
      </div>
      <div
        className="date"
      >
        Today at 8:39 PM
      </div>
    </div>
    <div
      className="details"
    >
      <div
        className="lastMessage"
      >
        Hello! How are you?
      </div>
    </div>
  </div>
</div>
`;

exports[`ChatItem with unread messages renders chat item 1`] = `
<div
  className="chatItem"
  onClick={[Function]}
>
  <div
    className="chatItemContainer"
  >
    <div
      className="interlocutorAvatar"
    >
      <img
        className="rounded-circle"
        height="32"
        src="/john_smith_avatar.jpg"
        width="32"
      />
    </div>
    <div
      className="details"
    >
      <div
        className="interlocutorFullName"
      >
        John Smith
      </div>
      <div
        className="date"
      >
        Today at 8:39 PM
      </div>
    </div>
    <div
      className="details"
    >
      <div
        className="lastMessage"
      >
        Hello! How are you?
      </div>
      <div
        className="badge badge badge-pill badge-secondary"
      >
        2
      </div>
    </div>
  </div>
</div>
`;
